lib/puppet/configurer.rb

```diff
@@ -296,14 +297,18 @@ def run_internal(options)
       configured_environment = Puppet[:environment] if Puppet.settings.set_by_config?(:environment)
 
       # We only need to find out the environment to run in if we don't already have a catalog
-      unless (cached_catalog || options[:catalog] || configured_environment)
-        Puppet.debug(_("No environment configured, attempting to find out the last used environment"))
-        if last_agent_environment
-          @environment = last_agent_environment
-          report.environment = last_agent_environment
+      unless (cached_catalog || options[:catalog] || Puppet.settings.set_by_cli?(:environment) || Puppet[:strict_environment_mode])
```

Comment:
If `Puppet[:strict_environment_mode]` is set, or if `environment` is set via CLI, there is no point in checking the lastrunfile; we just assume it's authoritative and fail further down if needed.